### ISSUE TITLE
GDScript lambda functions fail when a parameter is type-annotated with a nested class

### ISSUE THREAD

**Issue by Carnagion** [godotengine/godot#66259]
### Godot version

4.0 beta 1

### System information

Windows 10, Intel Core i7-10750H CPU, NVIDIA GeForce GTX 1650 Ti graphics, Vulkan

### Issue description

Refer to the **Steps to reproduce** section, and notice the `_ready()` function.
When run, its output is the following:
```
[<RefCounted#-9223372012343262836>]
[true]
[]
```

The first line - corresponding to `print(dictionary.values())` - is normal and expected. It simply prints out the dictionary's values as an array, and there is only one value (of type `Outer.Inner`, which happens to inherit `RefCounted` implicitly).

The second line - corresponding to `print(dictionary.values().map(func(inner): return inner is Outer.Inner))` - is also normal and expected. The `map()` function here maps each value to a boolean value - `true` if the value is of type `Outer.Inner` and `false` otherwise. The singluar value in the dictionary is definitely of type `Outer.Inner`, and so the result of `map()` is an array with a single `true` inside it.

The last line of the output - corresponding to `print(dictionary.values().map(func(inner: Outer.Inner) -> bool: return inner is Outer.Inner))` - is the issue. Notice that the lambda function in this `map()` is essentially the same as the previous one, but just with type annotations for the parameter and return type. Theoretically, this means it should return the same result as the previous `map()` - being a pure function. However, it returns an empty array instead. Examining the debugger output reveals two errors (expanded):
```
E 0:00:01:0236   _ready: Can't get method on CallableCustom "<anonymous lambda>(lambda)".
  <C++ Error>    Method/function failed. Returning: StringName()
  <C++ Source>   core/variant/callable.cpp:318 @ get_method()
  <Stack Trace>  main.gd:8 @ _ready()
E 0:00:01:0237   _ready: Error calling method from 'map': 'GDScript::': Cannot convert argument 1 from Object to Object.
  <C++ Error>    Method/function failed. Returning: Array()
  <C++ Source>   core/variant/array.cpp:472 @ map()
  <Stack Trace>  main.gd:8 @ _ready()
```

Notice that the latter error occurs because the engine "cannot convert argument 1 from Object to Object", which makes no sense, since an `Object` is already an `Object` and therefore no conversion is necessary (and any conversions should succeed all the time).

The errors disappear if the parameter's type annotation is removed, or if the nested class is moved to the same script file as the anonymous function, or if a regular (non-nested) class is used instead of a nested one.

**TL;DR:** Lambdas/anonymous functions do not execute as intended when one of their parameters is type-annotated with a nested class from a different script file.

### Steps to reproduce

Main script file:
```gdscript
extends Node

func _ready():
    var dictionary: Dictionary = {}
    dictionary[1] = Outer.Inner.new()
    print(dictionary.values())
    print(dictionary.values().map(func(inner): return inner is Outer.Inner))
    print(dictionary.values().map(func(inner: Outer.Inner) -> bool: return inner is Outer.Inner))
```
In another script file:
```gdscript
class_name Outer

class Inner:
    
    func _init():
        pass
```
Run the main script file by attaching it to a main scene and examine the standard and error output.

### Minimal reproduction project

As there are only two steps to reproduce the issue, a minimal reproduction project has not been included.

**Comment by bsil78** [comment]
@Carnagion  : I cannot reproduce in Godot beta 7 ; it seems fixed, no ?

**Comment by adamscott** [comment]
Could not reproduce on master (fcba87e). @akien-mga We can close this issue.
